### PR TITLE
Support extra state in buildbot.process.properties.renderer

### PR DESCRIPTION
**Pull request by p12tic**
Currently `buildbot.process.properties.renderer` supports only simple renderings that are dependent only on the properties passed to it. In more complex buildbot configurations where large part of it is generated it's sometimes needed to have a renderer that has access to full properties set of the build and also depends on some other data which could vary. Unfortunately, `Transform` does not satisfy this use case because it only allows predefined properties to be combined and does not give both properties and extra arguments to the same function. Thus the only solution is to implement custom renderer that satisfies the `IRenderable` interface and stores that extra data as part of state.

I propose that we allow supplying extra arguments to the object returned by `buildbot.process.properties.renderer` decorator that are later passed to the renderer function. For example:

```
@util.renderer
def my_test_renderer(props, extra_arg, extra_kwarg='default'):
    return '%s-%s-%s' % (props.getProperty('myprop'), extra_arg, extra_kwarg)

...

SomeStep(some_renderable_arg=my_test_renderer.withArgs('arg', extra_kwarg='kw'), ...)
```

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragment` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

**Comment by p12tic** [comment]
Looks like pywin is not install for some reason on Appveyor CI server.

File: master/buildbot/process/properties.py [hunk 1]
```diff
@@ -897,14 +897,30 @@ class _Renderer(util.ComparableMixin, object):
...
+    @defer.inlineCallbacks
     def getRenderingFor(self, props):
+        args = yield props.render(self.args)
+        kwargs = yield props.render(self.kwargs)
+
         # We allow the renderer fn to return a renderable for convenience
-        d = defer.maybeDeferred(self.fn, props)
-        d.addCallback(props.render)
-        return d
+        result = yield defer.maybeDeferred(self.fn, props, *args, **kwargs)
```
Comment (tardyp) [inline master/buildbot/process/properties.py]: no need for maybeDeferred if the result is yield in inlineCallbacks

```
        result = yield self.fn(props, *args, **kwargs)
```

yield accepts both deferred or non deferred result.

File: master/buildbot/process/properties.py [hunk 2]
```diff
@@ -897,14 +897,30 @@ class _Renderer(util.ComparableMixin, object):
...
+        result = yield props.render(result)
+        defer.returnValue(result)
 
     def __repr__(self):
+        if self.args or self.kwargs:
+            return 'renderer(%r, args=%s, kwargs=%s)' % (self.fn, self.args,
```
Comment (tardyp) [inline master/buildbot/process/properties.py]: I would rather use %r for args and kwargs

File: master/buildbot/test/unit/test_process_properties.py [hunk 3]
```diff
@@ -1366,53 +1366,56 @@ def setUp(self):
...
 
+    @defer.inlineCallbacks
     def test_renderer(self):
         self.props.setProperty("x", "X", "test")
-        d = self.build.render(
-            renderer(lambda p: 'x%sx' % p.getProperty('x')))
-        d.addCallback(self.assertEqual, 'xXx')
-        return d
 
-    def test_renderer_called(self):
```
Comment (tardyp) [inline master/buildbot/test/unit/test_process_properties.py]: Why did you remove this test? looks like an overlook.

File: master/buildbot/process/properties.py (hunk 1 above)
Comment (p12tic) [reply master/buildbot/process/properties.py]: Fixed now, thanks!

File: master/buildbot/process/properties.py (hunk 2 above)
Comment (p12tic) [reply master/buildbot/process/properties.py]: Indeed. Fixed now.

File: master/buildbot/test/unit/test_process_properties.py (hunk 3 above)
Comment (p12tic) [reply master/buildbot/test/unit/test_process_properties.py]: Yeah, the first proposal that I had in mind was making the renderable callable instead of adding the `withArgs` method. I've removed the test and forgot the add it after figuring out that the callable API would be completely unreadable and unexpected. Thanks for noticing this! Fixed now.
